src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/MacProxy.cs

```diff
@@ -0,0 +1,125 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Net.Http;
+using System.Net;
+using System.Collections.Generic;
+using Microsoft.Win32.SafeHandles;
+using static Interop.CoreFoundation;
+using static Interop.RunLoop;
+
+using CFRunLoopRef = System.IntPtr;
+
+namespace System.Net.Http
+{
+    internal sealed class MacProxy : IWebProxy
+    {
+        public ICredentials Credentials
+        {
+            get => null;
+            set => throw new NotSupportedException();
+        }
+
+        static Uri GetProxyUri(string scheme, CFProxy proxy)
+        {
+            var uriBuilder = new UriBuilder(
+                scheme,
+                proxy.HostName,
+                proxy.PortNumber);
+
+            // FIXME: Handle user name and password
+
+            return uriBuilder.Uri;
+        }
+
+        public Uri ExecuteProxyAutoConfiguration(SafeCreateHandle cfurl, CFProxy proxy)
+        {
+            Uri result = null;
+            CFRunLoopRef runLoop = CFRunLoopGetCurrent();
+
+            // Callback that will be called after executing the configuration script
+            CFProxyAutoConfigurationResultCallback cb = (IntPtr client, IntPtr proxyListHandle, IntPtr error) =>
```

Comment:
We have a similar code running in liveshare, the source code ported from mono. We have been getting reports from Darwin 17.7.0 that the callback gets garbage collected, looks like the unmanaged code is not holding the reference to this. Although it is in scope of the local variable, I'm not sure how this happens. The code pattern here is exactly the same as we have and can have the same issue.

```
FailFast:
A callback was made on a garbage collected delegate of type 'vsls-agent!Microsoft.Cascade.Platforms.Mac.CFNetwork+CFProxyAutoConfigurationResultCallback::Invoke'.​
[2019-02-21 00:23:59.222 Client.Agent I] at Microsoft.Cascade.Platforms.Mac.CFRunLoop.CFRunLoopRunInMode(IntPtr, Double, Boolean)​
   at Microsoft.Cascade.Platforms.Mac.CFRunLoop.CFRunLoopRunInMode(IntPtr, Double, Boolean)​
   at Microsoft.Cascade.Platforms.Mac.CFNetwork.ExecuteProxyAutoConfigurationURL(IntPtr, System.Uri)​
   at Microsoft.Cascade.Platforms.Mac.CFWebProxy.GetProxy(System.Uri)​
[2019-02-21 00:23:59.222 Client.Agent I] at Microsoft.Cascade.Platforms.Mac.CFWebProxy.IsBypassed(System.Uri)​
   at System.Net.Http.HttpConnectionPoolManager.SendAsync(System.Net.Http.HttpRequestMessage, Boolean, System.Threading.CancellationToken)​
   at System.Net.Http.HttpConnectionHandler.SendAsync(System.Net.Http.HttpRequestMessage, System.Threading.CancellationToken)​
   at System.Net.Http.RedirectHandler+<SendAsync>d__4.MoveNext()​
   at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[[System.Net.Http.RedirectHandler+<SendAsync>d__4, System.Net.Http, Version=4.2.1.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a]](<SendAsync>d__4 ByRef)​
   at System.Net.Http.RedirectHandler.SendAsync(System.Net.Http.HttpRequestMessage, System.Threading.CancellationToken)​
   at System.Net.Http.SocketsHttpHandler.SendAsync(System.Net.Http.HttpRequestMessage, System.Threading.CancellationToken)​
   at System.Net.Http.HttpMessageInvoker.SendAsync(System.Net.Http.HttpRequestMessage, System.Threading.CancellationToken)​
```

Reply:
Thanks, it makes sense to explicitly call `GC.KeepAlive`. I will add it.

Note that there are two other issues for the Mono code that you may have inherited.

Firstly, the Mono code does not call `CFRunLoopSourceInvalidate(loopSource)`. This can probably result in errors in some obscure scenarios. There were issues with it ~4 years ago in Chromium and they fixed it, so perhaps it will be possible to find some context there.

Secondly, the Mono code for autoconfiguration scripts (not URLs) uses different API than I use. That already lead to issues with multi-threaded code before and resulted in some obscure fix with always running the query in the same thread. I went through the history of the issue and I believe my approach with `CFNetworkExecuteProxyAutoConfigurationScript` is a more correct solution with way less code. The original issue was happening on iOS 6, so there's no way for me to check whether the problem would still happen on that platform. I believe the API I use actually runs the JavaScript out of process, but I couldn't find any documentation explicitly stating it.